ipcore/vhdl_node.py

```diff
@@ -0,0 +1,373 @@
+        self.libraries = libraries or []
+    @abstractmethod
+    def generate(self):
+        """Generate VHDL code."""
+    def write(self, output_dir=""):
+        """Write file to the filesystem.
+
+        Copies packages used in the work library to the output directory as well.
+
+        Parameters
+        ----------
+        output_dir : str, optional
+            Output directory to write file to, by default "" which results in the current directory
+        """
+        if output_dir != "" and not output_dir.endswith(os.sep):
+            output_dir += os.sep
+
+        with open(output_dir + self.name + ".vhd", 'w') as vhdl_file:
+            vhdl_file.write(self.generate())
+        
+        work_library = next((library for library in self.libraries if library.name == "work"), None)
+        cur_dir = pathlib.Path(os.path.realpath(__file__)).parent.name
+
+        if work_library is not None:
+            for package in work_library.packages:
+                if os.path.isfile(cur_dir + "/" + package + ".vhd"):
+                    copyfile(cur_dir + "/" + package + ".vhd", output_dir + package + ".vhd")
+
+class EntityFile(VHDLFile):
+    """Represents a VHDL entity file."""
+
+    def __init__(self, name, entity, architecture=None, libraries=None):
+        super().__init__(name, libraries)
+        self.entity = entity
+        self.architecture = architecture
+    def generate(self):
+        """Generate a VHDL entity file."""
+        vhdl = ""
+
+        for library in self.libraries:
+            vhdl += library.generate()
+
+        vhdl += self.entity.generate()
+        vhdl += self.architecture.generate()
+        return vhdl
+
+class Signal(BaseVHDLNode):
+    def __init__(self, name, length=1, default_value=None, data_type=None, underlying_data_type=None):
+        super().__init__(name)
+        self.length = length
+        self.default_value = default_value
+
+        if data_type is None:
+            data_type = self._get_data_type()
+        self.data_type = data_type
+        self.underlying_data_type = underlying_data_type
+    def generate(self):
+        """Generate a VHDL signal."""
+        if self.length == 1:
+            type_declaration = self.data_type
+        else:
+            type_declaration = f"{self.data_type}({(self.length - 1)} downto 0)"
+        if self.default_value is not None:
+            default_str = f" := {self._get_default_str()}"
+        else:
+            default_str = ""
+        return f"signal {self.name.ljust(32)} : {type_declaration}{default_str};\n"
+    def generate_assignment(self, new_signal_value):
+        """Generate VHDL assignment to this signal, including necessary resizing.
+
+        Resizes using underlying data type if possible.
+        VHDL data types must match except for std_logic and std_logic_vector.
+        For assignmnents between std_logic and std_logic_vector,
+        the first bit is where data is assumed to be stored.
```

Comment:
I mean, that isn't really true since it is only strictly the 0th bit (which I feel like is fair to equate with the first bit). If someone did `std_logic_vector(0 to 5)` then the 0th bit is the most significant bit. While I always use `downto`, I don't ever check for that when doing the single bit special case.

Reply:
True, the 0th bit is only the LSB when using `downto`. Darn semantics :unamused: